build.rs

```diff
@@ -1,116 +1,88 @@
-        sdk_output_dir
-            .join("include")
-            .join("c_api")
-            .to_str()
-            .unwrap(),
-    );
-
-    println!("cargo:rustc-link-search=native={}", lib_path.display());
-    println!("cargo:rustc-link-lib=dylib=Microsoft.CognitiveServices.Speech.core");
-
-    let skip_bindgen = env::var("MS_COG_SVC_SPEECH_SKIP_BINDGEN")
-        .map(|v| v == "1")
-        .unwrap_or(false);
-
-    if skip_bindgen {
+    let link_params = if let Some(lib_dir) = env_var("MS_COG_SVC_SPEECH_LIB_DIR") {
+        // If the user has specified an SDK lib directory then use that and
+        // don't try and download another one.
+        let include_dir = env_var("MS_COG_SVC_SPEECH_INCLUDE_DIR").map(PathBuf::from);
+        match TargetOs::get() {
+            TargetOs::Linux => LinkParams {
+                lib_arg: "dylib=Microsoft.CognitiveServices.Speech.core".to_owned(),
+                search_arg: format!("native={}", lib_dir.to_str().unwrap()),
+                include_dir,
+            },
+            TargetOs::MacOS => LinkParams {
+                lib_arg: "framework=MicrosoftCognitiveServicesSpeech".to_owned(),
+                search_arg: format!("framework={}", lib_dir.to_str().unwrap()),
+                include_dir,
+            },
+            TargetOs::Windows => LinkParams {
+                lib_arg: "dylib=Microsoft.CognitiveServices.Speech.core".to_owned(),
+                search_arg: format!("native={}", lib_dir.to_str().unwrap()),
+                include_dir,
+            },
+        }
+    } else {
+        let download_dir = {
+            let mut dir = PathBuf::from(std::env::var_os("OUT_DIR").unwrap());
+            dir.push("sdk_output");
+            dir
+        };
+        // Don't bother re-downloading the SDK if it looks like it's already
+        // there, unless it's been explicitly requested.
+        if env_var("MS_COG_SVC_SPEECH_RENEW_SDK").is_some()
+            || !download_dir.exists()
+            || fs::read_dir(&download_dir).unwrap().next().is_none()
+        {
+            fs::create_dir_all(&download_dir).unwrap();
+            download_sdk(&download_dir);
+        }
+        let sdk_dir = match TargetOs::get() {
+            TargetOs::Linux => download_dir.join(format!("SpeechSDK-Linux-{}", SPEECH_SDK_VERSION)),
+            TargetOs::MacOS => download_dir.join("MicrosoftCognitiveServicesSpeech.xcframework"),
+            TargetOs::Windows => download_dir,
+        };
+        link_params(&sdk_dir)
+    };
+
+    println!("cargo:rustc-link-search={}", link_params.search_arg);
+    println!("cargo:rustc-link-lib={}", link_params.lib_arg);
+
+    if env_var("MS_COG_SVC_SPEECH_SKIP_BINDGEN").is_some() {
         return;
     }
+    println!("cargo::rustc-cfg=bindgen");
 
-    // The bindgen::Builder is the main entry point
-    // to bindgen, and lets you build up options for
-    // the resulting bindings.
+    let Some(include_dir) = link_params.include_dir else {
```

Comment:
not sure but could this be done into ` let Some(_) = link_params.include_dir else {` ?

or maybe 

```
if  link_params.include_dir.is_none() {
panic!...
}

```

I was fighting little bit to read it

Reply:
I'm sorry, I don't quite understand what you're asking for. Do you not like the (new-ish) let-else syntax? I could convert it into a match statement if you would be happier with that?
```rust
let include_dir = match link_params.include_dir {
    Some(include_dir) => include_dir,
    _ => {
        // Note this is only possible to hit if the user has specified the lib dir.
        panic!("Include directory must be specified with MS_COG_SVC_SPEECH_INCLUDE_DIR");
    }
};
```

Reply:
got it now, but I really don't like it. Match (as suggested above) would be in this case much better.

Reply:
Ok, replaced with the match statement.